### PR TITLE
Do dependencies in 1 command per file class.

### LINKED ISSUE #5988: coqdep is slow (coqdep should have `-o`, and coq_makefile should use it)
In large projects (e.g., [fiat-crypto](https://github.com/mit-plv/fiat-crypto)), `coqdep` takes many minutes to complete on a fresh checkout.  On Gitter, @herbelin suggested giving `coqdep` a `-o` option to take a list of files on the command line, and output all of the `.v.d` files.  Some testing suggests this would be a massive improvement.  Timing results on linux:
```bash
$ coqtop -v
The Coq Proof Assistant, version 8.7.0 (October 2017)
compiled on Oct 18 2017 11:37:05 with OCaml 4.02.3
$ uname -a
Linux jgross-Leopard-WS 4.4.0-66-generic #87-Ubuntu SMP Fri Mar 3 15:29:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.2 LTS
Release:        16.04
Codename:       xenial
$ git ls-files "src/*.v" | wc -l
3617
$ # coqdep, all at once, no file writing
$ time (export COQPATH="$(cd coqprime && pwd)"; git ls-files "src/*.v" | xargs coqdep -R src Crypto -R Bedrock Bedrock >/dev/null)

real    0m0.904s
user    0m0.772s
sys     0m0.108s
$ # coqdep, all at once, writing to a file
$ time (export COQPATH="$(cd coqprime && pwd)"; git ls-files "src/*.v" | xargs coqdep -R src Crypto -R Bedrock Bedrock >tmp)

real    0m1.039s
user    0m0.744s
sys     0m0.152s
$ # coqdep, one at a time, all in parallel
$ time (export COQPATH="$(cd coqprime && pwd)"; for i in $(git ls-files "src/*.v"); do (coqdep -R src Crypto -R Bedrock Bedrock "$i" >/dev/null) & echo "$i" ; done)
...
real    1m21.849s
user    2m32.604s
sys     1m10.772s
$ # just the for loop
$ time (export COQPATH="$(cd coqprime && pwd)"; for i in $(git ls-files "src/*.v"); do echo "$i" >/dev/null ; done)

real    0m0.054s
user    0m0.024s
sys     0m0.012s
$ # just the for loop, and echoing
$ time (export COQPATH="$(cd coqprime && pwd)"; for i in $(git ls-files "src/*.v"); do echo "$i" ; done)
...
real    0m0.090s
user    0m0.024s
sys     0m0.000s
$ # coqdep, one at a time, in serial
$ time (export COQPATH="$(cd coqprime && pwd)"; for i in $(git ls-files "src/*.v"); do coqdep -R src Crypto -R Bedrock Bedrock "$i" >/dev/null; done)

real    6m28.858s
user    2m44.356s
sys     1m16.668s
```

### PR DESCRIPTION
**Pull request by SkySkimmer**
coq_makefile produces a Makefile which does `coqdep $(FLAGS) $(VFILES) > .vfiles.d`
Coq's own Makefiles also produce coalesced `.mlfoo.d` ([ml/mli use ocamldep + mllib/mlpack use ocamllibdep] * [checker / plugins / rest use separate flags]). This wasn't done for coq_makefile but could be if there's demand.

Unimath is broken because they check that the order of tag generation is compatible with the dependency order, which they use the v.d files for.

Fix #5988, see benchmark https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/241/console (with caveat that coqdep might still be doing inefficient things ofc)
```
┌──────────────────────────┬──────────────────────────┬───────────────────────────────────────┬────────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                          │      user time [s]       │              CPU cycles               │            CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                          │                          │                                       │                                        │                         │                 │
│             package_name │     NEW     OLD  PDIFF   │           NEW            OLD  PDIFF   │            NEW            OLD  PDIFF   │     NEW     OLD PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-core │  100.60  119.70 -15.96 % │  279032379770   315751328583 -11.63 % │   363445071048   424116612776 -14.31 % │  502232  502572 -0.07 % │  15   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│          coq-fiat-crypto │ 3356.62 3716.68  -9.69 % │ 9355596792764 10341719205058  -9.54 % │ 15836566246788 17539300025959  -9.71 % │ 3282944 3282676 +0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  651.73  675.51  -3.52 % │ 1821538306849  1867845688099  -2.48 % │  2966511358622  3029504433708  -2.08 % │ 3535648 3534884 +0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-odd_order │ 1362.21 1382.99  -1.50 % │ 3792155295519  3855828453697  -1.65 % │  6704276847509  6703997409949  +0.00 % │ 1121452 1109976 +1.03 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-geocoq │ 3056.72 3098.26  -1.34 % │ 8530157579011  8645086519808  -1.33 % │ 14290179855016 14317259972179  -0.19 % │ 1216268 1250012 -2.70 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-sf-vfa │   18.56   18.74  -0.96 % │   51833037991    51992857232  -0.31 % │    70938041264    71003227643  -0.09 % │  537372  537420 -0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   59.07   59.64  -0.96 % │  163187831745   164356618411  -0.71 % │   229166960267   229639694955  -0.21 % │  530916  531244 -0.06 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  267.10  269.26  -0.80 % │  742950126451   748274284417  -0.71 % │  1100695927681  1100851669853  -0.01 % │  981588  981696 -0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   38.05   38.34  -0.76 % │  105250679200   105625546166  -0.35 % │   130829168640   131452165963  -0.47 % │  477592  477404 +0.04 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-color │  566.43  570.58  -0.73 % │ 1586302193738  1595581304010  -0.58 % │  2004234150999  2015849829353  -0.58 % │ 1488188 1487940 +0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  448.65  451.86  -0.71 % │ 1247819523502  1258024855141  -0.81 % │  2095874074865  2096406956637  -0.03 % │  729044  728224 +0.11 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  168.54  169.29  -0.44 % │  468733459198   471154710670  -0.51 % │   668518826832   668781608887  -0.04 % │  590200  590224 -0.00 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-mathcomp-real_closed │  166.24  166.94  -0.42 % │  462229342259   464153163350  -0.41 % │   711970726247   712398950777  -0.06 % │  731516  732284 -0.10 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-vst │ 2177.01 2185.99  -0.41 % │ 6071022822166  6093105493201  -0.36 % │  8896882744656  8891900364981  +0.06 % │ 1951524 1951828 -0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  189.20  189.92  -0.38 % │  525885860360   528263563838  -0.45 % │   768026202487   768220029033  -0.03 % │  715068  714936 +0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│           coq-coquelicot │   74.66   74.73  -0.09 % │  205598006569   205687035528  -0.04 % │   258842413511   258920649977  -0.03 % │  659968  660056 -0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-sf-plf │   41.84   41.87  -0.07 % │  116552940203   116460303871  +0.08 % │   153002937171   153055790679  -0.03 % │  510672  510680 -0.00 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-bignums │   75.00   75.04  -0.05 % │  207726428237   208795928383  -0.51 % │   284421738640   285375454180  -0.33 % │  535420  535280 +0.03 % │  24   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-flocq │   58.04   57.95  +0.16 % │  159054344225   158975740972  +0.05 % │   205329186602   205373779533  -0.02 % │  643768  643876 -0.02 % │ 161   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-compcert │  802.95  801.36  +0.20 % │ 2239706838770  2234838124321  +0.22 % │  3413311591191  3415512000564  -0.06 % │ 1387948 1387876 +0.01 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-formal-topology │   33.84   33.66  +0.53 % │   93397098561    94179630472  -0.83 % │   125100821458   126873449235  -1.40 % │  482640  482564 +0.02 % │   0   0  +nan % │
├──────────────────────────┼──────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-sf-lf │   15.54   15.43  +0.71 % │   42429693854    42418332415  +0.03 % │    59550787458    59568498344  -0.03 % │  419096  419124 -0.01 % │   0   0  +nan % │
└──────────────────────────┴──────────────────────────┴───────────────────────────────────────┴────────────────────────────────────────┴─────────────────────────┴─────────────────┘
```

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Fiat is supposed to be faster, but it timed out on travis...

**Comment by SkySkimmer** [comment]
Getting some error on fiat https://travis-ci.org/SkySkimmer/coq/jobs/308384592
```
File "./src/Specific/NISTP256/AMD128/Synthesis.v", line 6, characters 9-31:

Error:

In nested Ltac calls to "make_Synthesis_package", "get_Synthesis_package",

"add_Synthesis_package",

"MontgomeryReificationTypesPackage.add_MontgomeryReificationTypes_package",

"MontgomeryReificationTypesPackage.add_feBW_tight_of_feBW_small",

"CurveParametersPackage.if_montgomery", "tac_true" (bound to

fun _ =>

  let feBW_tight :=

   TagList.Tag.get pkg ReificationTypesPackage.TAG.feBW_tight

  in

  let feBW_small :=

   TagList.Tag.get pkg MontgomeryReificationTypesPackage.TAG.feBW_small

  in

  let feBW_tight_of_feBW_small := fresh "feBW_tight_of_feBW_small" in

  let feBW_tight_of_feBW_small :=

   MontgomeryReificationTypes.pose_feBW_tight_of_feBW_small feBW_tight

    feBW_small feBW_tight_of_feBW_small

  in

  TagList.Tag.update pkg

   MontgomeryReificationTypesPackage.TAG.feBW_tight_of_feBW_small

   feBW_tight_of_feBW_small),

"MontgomeryReificationTypes.pose_feBW_tight_of_feBW_small",

"CacheTerm.cache_term_with_type_by", "CacheTerm.cache_term_with_type_by_gen",

"abstract_tac" (bound to fun id' => transparent_abstract tac using id') and

"transparent_abstract (tactic3) using (ident)", last call failed.

Illegal application: 

The term "proj1_sig" of type

 "forall (A : Type) (P : A -> Prop), {x : A | P x} -> A"

cannot be applied to the terms

 "feBW_tight" : "Type"

 "fun v : feBW_tight =>

  BinInt.Z.lt (meval v) (MontgomeryAPI.eval (BinNums.Zpos r) m_enc)"

   : "feBW_tight -> Prop"

The 1st term has type "Type@{Crypto.Specific.NISTP256.AMD128.Synthesis.16}"

which should be coercible to "Type@{Coq.Init.Specif.16}".
```

**Comment by JasonGross** [comment]
Oh, er, this is because fiat-crypto's lite target uses [makefile magic and sed scripts](https://github.com/JasonGross/coq-scripts/blob/master/Makefile.vo_closure) to determine the transitive dependency closure of a set of `.vo` files by parsing the `.v.d` files; we list some files to exclude, and compute the set of files that depend on them, to exclude them all.  Would it be possible for `coqdep` to get a mode that says "given this list of files, compute the set that depends on this particular file, transitively" and a mode that says "given this list of files, compute the full set that this file depends on, transitively"?

If not, either @andersk (the original author of the makefile code for `vo_closure`) or I should be able to modify it to handle having only a single `.d` file, probably cased on the existence of the `VDFILE` variable...

**Comment by JasonGross** [comment]
@SkySkimmer Can you check if this PR works if coq-scripts gets the diff (untested) on `Makefile.vo_closure`:
```diff
# with the deduplicated newly found dependencies.  When $1 becomes
# empty, the result is $2.
+ifdef VDFILE
+# Coq >= 8.8
+read_deps = $(if $(wildcard $(VDFILE).d),$(filter %$(1:.v.d=.vo),$(shell sed -n 's/^[^:]*: // p' $(wildcard $(VDFILE).d))))
+else
read_deps = $(if $(wildcard $1),$(filter %.vo,$(shell sed -n 's/^[^:]*: // p' $(wildcard $1))))
+endif
vo_closure = $(if $1,$(call vo_closure,$(sort $(filter-out $1 $2,$(call read_deps,$(1:.vo=.v.d)))),$1 $2),$2)
```

I'm not at a computer with easy access to building Coq right now, but I can also try to test this later today.

**Comment by SkySkimmer** [comment]
https://travis-ci.org/SkySkimmer/coq/jobs/310104242 started

**Comment by SkySkimmer** [comment]
Doesn't seem to work.

**Comment by JasonGross** [comment]
@SkySkimmer You updated the path in `.gitmodules` but didn't update the commit that it points to; https://github.com/SkySkimmer/fiat-crypto/tree/master/etc says `coq-scripts @ 617ccff` when it should be pointing at `ff2d1bf`

**Comment by SkySkimmer** [comment]
I blame git.
Updated and job rebooted.

**Comment by SkySkimmer** [comment]
Still not working.

**Comment by JasonGross** [comment]
oh, duh, I was filtering on the wrong .vo's.  Try this (still haven't had a chance to try it myself, alas):
```diff
# with the deduplicated newly found dependencies.  When $1 becomes
# empty, the result is $2.
+ifdef VDFILE
+# Coq >= 8.8
+read_deps = $(if $(wildcard $(VDFILE).d),$(filter %.vo,$(shell grep $(1:.v.d=.vo) $(wildcard $(VDFILE).d) | sed -n 's/^[^:]*: // p')))
+else
read_deps = $(if $(wildcard $1),$(filter %.vo,$(shell sed -n 's/^[^:]*: // p' $(wildcard $1))))
+endif
vo_closure = $(if $1,$(call vo_closure,$(sort $(filter-out $1 $2,$(call read_deps,$(1:.vo=.v.d)))),$1 $2),$2)
```

**Comment by SkySkimmer** [comment]
Now it fails by nontermination immediately after generating the VDFILE.d
```make
-*- mode: compilation; default-directory: "~/dev/coq/fiat-crypto/" -*-
Compilation started at Fri Dec  1 22:11:52

make lite
COQ_MAKEFILE -f _CoqProject > Makefile.coq
COQDEP VFILES
*** Warning: in file src/Util/NumTheoryUtil.v, library Coqprime.Zp is required and has not been found in the loadpath!
*** Warning: in file src/Util/NumTheoryUtil.v, library Coqprime.Zp is required and has not been found in the loadpath!

Compilation interrupt at Fri Dec  1 22:12:32
```

**Comment by JasonGross** [comment]
Oh, er, do we need to `grep` for `^$(1:.v.d=.vo)` or for `$(1:.v.d=.vo):` or something?  (How to pick out the rules that are for this .vo file, rather than the rules where this vo file shows up)  I'm just guessing here, though.  I'll try to work on this myself in the next few days, if you don't figure it out before then.

**Comment by JasonGross** [comment]
@SkySkimmer I've pushed an update to coq-scripts and fiat-crypto master that should make this work fine now.  (In addition to what I suggested above, I also needed to separate out the calls finding the deps for each vo file, to make the simple test I was using (making sure the output of `make printdeps src/Demo.vo` stayed the same) passed.)

**Comment by SkySkimmer** [comment]
Seems to work https://travis-ci.org/SkySkimmer/coq/jobs/313597841

File: Makefile.build [hunk 1]
```diff
@@ -670,21 +676,24 @@ plugins/%.cmx: plugins/%.ml
...
+MAINMLLIBFILES := $(filter-out checker/%, $(filter-out plugins/%, $(MLLIBFILES) $(MLPACKFILES)))
+
+$(MLDFILE).d: $(D_DEPEND_BEFORE_SRC) $(MAINMLFILES) $(D_DEPEND_AFTER_SRC) $(GENFILES)
+	$(SHOW)'OCAMLDEP  MLFILES MLIFILES'
+	$(HIDE)$(OCAMLDEP) $(DEPFLAGS) $(MAINMLFILES) $(TOTARGET)
 
-%.mli.d: $(D_DEPEND_BEFORE_SRC) %.mli $(D_DEPEND_AFTER_SRC) $(GENFILES)
-	$(SHOW)'OCAMLDEP  $<'
-	$(HIDE)$(OCAMLDEP) $(DEPFLAGS) "$<" $(TOTARGET)
+$(MLLIBDFILE).d: $(D_DEPEND_BEFORE_SRC) $(MAINMLLIBFILES) $(D_DEPEND_AFTER_SRC) $(OCAMLLIBDEP) $(GENFILES)
```
Comment (JasonGross) [inline Makefile.build]: I am not sure that `D_DEPEND_BEFORE_SRC` makes sense anymore with this change (removing `$<`).  @gares ?

**Review by JasonGross — APPROVED** [review]
Minor comments on Coq's Makefile.  I'm strongly in favor of getting this in soon.

**Comment by SkySkimmer** [comment]
Do we want to keep a mode where we output many small files?

**Comment by JasonGross** [comment]
> Do we want to keep a mode where we output many small files?

Seems like it would be a bit complicated, and I don't see a use case for it, but I won't object to it.

**Comment by SkySkimmer** [comment]
>Seems like it would be a bit complicated, and I don't see a use case for it, but I won't object to it.

It would probably be feasible to put the changes of this PR behind `ifndef MANY_SMALL_FILES` if annoying to keep it well tested, but I'm not doing it without someone wanting it.

**Comment by ejgallego** [comment]
An interesting effect of this PR is that if I modify a single `ml` file, now `ocamldep` is called again on all the `.ml` files as opposed to the single file call used before.

Not a big deal IMHO (~1 s) but I thought I would document it.

**Pull request merged**